Update trigger threshold guess estimation to mean of trigger channel

62442katieb opened the pull request:
<!-- Write all of the issues that are linked to this pull request. -->
<!-- If this PR is enough to close them you can write something like "Closes #314 and closes #42" -->
<!-- If you just want to reference them without closing them, you can add something like "References #112" -->
Closes #211 

<!-- Add a short description of the PR content here-->


## Proposed Changes
<!-- List major points of changes here, so that the reviewers can have a bit more context while looking at your work! -->
  - Trigger threshold is now defined as the mean value of the trigger channel across time
  - Logic: if trigger channel is binary (or roughly binary), the "on" state is always greater than the mean and the "off" state is always less than the mean

## Change Type
<!-- Indicate the type of change you think your pull request is -->
- [x] `bugfix` (+0.0.1)
- [ ] `minor` (+0.1.0)
- [ ] `major`  (+1.0.0)
- [ ] `refactoring` (no version update)
- [ ] `test` (no version update)
- [ ] `infrastructure` (no version update)
- [ ] `documentation` (no version update)
- [ ] `other`

## Checklist before review
<!-- If this section is not clear, please read this part of the docs: https://phys2bids.readthedocs.io/en/latest/contributorfile.html#pr -->
<!-- You're invited to open a draft PR ASAP, but before marking it "ready for review", check that you have done the following: -->
- [x] I added everything I wanted to add to this PR.
- [x] \[Code or tests only\] I wrote/updated the necessary docstrings.
- [ ] \[Code or tests only\] I ran and passed tests locally.
- [ ] \[Documentation only\] I built the docs locally.
- [x] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
- [x] My code respects the adopted style, especially linting conventions.
- [x] The title of this PR is explanatory on its own, enough to be understood as part of a changelog.
- [x] I added or indicated the right labels.
<!-- If relevant, you can add a milestone label or indicate an ideal timeline for this PR, including whether the progress of this PR is linked to other PRs being completed before or after it. -->
- [ ] I added information regarding the timeline of completion for this PR.
<!-- If you want, you can ask for reviews on a draft PR -->
- [ ] Please, comment on my PR while it's a draft and give me feedback on the development!

vinferrer commented on phys2bids/physio_obj.py:
```diff
@@ -467,11 +468,15 @@ def check_trigger_amount(self, thr=None, num_timepoints_expected=0, tr=0):
         if thr is None:
-            thr = np.mean(trigger) + 2 * np.std(trigger)
-            flag = 1
+            thr = np.mean(trigger) + np.std(trigger)	            # If trigger channels are binary 
```
If the threshold is gonna be the mean of the trigger, this line is not used at all then. Could you delete it?

62442katieb replied:
Oops! Yes, will do. that was a copy/paste error.

vinferrer reviewed (changes requested):
Also you should look into the docs and update them, For example: [Here in line 205](https://github.com/62442katieb/phys2bids/blob/bug/trigger-thresh/docs/howto.rst ). There is also a linting error, seems like you left white spaces at the end of some lines. After those changes, it LGTM for approval

62442katieb commented:
> Also you should look into the docs and update them, For example: [Here in line 205](https://github.com/62442katieb/phys2bids/blob/bug/trigger-thresh/docs/howto.rst). 
Good catch, fixed! 👍 

>There is also a linting error, seems like you left white spaces at the end of some lines. After those changes, it LGTM for approval
Linting errors also fixed 👍

vinferrer commented:
LGTM!! @smoia do you think this is a Bugfix? in that case we don't need a second review

vinferrer reviewed (approved):
LGTM

smoia reviewed (commented):
I'd say it 's LeGiTiMate, but I'm probably over-worrying about very long multi-run recordings where you have a lot of time between runs and the mean might be too sensitive to noise. 

May I ask @sangfrois or @RayStick or anyone else that has a good, long multi-run session with a noisy trigger channel at disposal to run this PR and communicate the results?

RayStick commented:
I'm happy to do some testing (but I can only do that Tues 26th, in case there is a specific timeline needed)

smoia commented:
No timeline, go for it!

RayStick commented:
I'm having trouble with my python/phys2bids set-up at the moment - sorry! I can check this again on Wednesday to see if i can troubleshoot it but the data I was going to test on OSF**: labchart/chicago/four_scans_samefreq_allchannels.txt** and has a json file.

sangfrois commented:
Hey @RayStick ! No problem, I got this.
I tested it with a multi-run acknowledge file. It has 8 runs. It spans about 1h30. 


### The trigger signal looks like this 
**Note that Y axis is Voltage**
![image](https://user-images.githubusercontent.com/38385719/115892420-210a4b00-a425-11eb-9cb8-749eea311093.png)
### The trigger actually looks like this
![image](https://user-images.githubusercontent.com/38385719/115890456-f1f2da00-a422-11eb-8c83-11ed5a919cd7.png)
![image](https://user-images.githubusercontent.com/38385719/115930859-2cc33500-a458-11eb-8eda-4c3cc2da6254.png)
@smoia the noise seems to be trivial. So I wouldn't be worried to derive a threshold based on the mean that would confound noise and real triggers.

### The given input is
``phys2bids -in path/to/multi-run.acq -ntp 406 406 406 3 410 410 410 2 -tr 1.49 -chtrig 4``
### The output is
```
INFO       Currently running phys2bids version 2.2.3+725.g3b4ec90.dirty
INFO       Input file is C:/Users/sangfrois/Desktop/NeuroMod_tests/multi-run.acq
INFO       File extension is .acq
WARNING    If both acq and txt files exist in the path, acq will be selected.
INFO       Reading the file C:\Users\sangfrois\Desktop\NeuroMod_tests\multi-run.acq
INFO       00. Channel GSR-EDA100C-MRI: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       01. Channel PPG100C: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       02. Channel ECG100C: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       03. Channel TTL: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       04. Channel Custom, HLT100C - A 6: 26212257 samples, 5000.0 samples/sec, loaded: True
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Checking that units of measure are BIDS compatible
INFO       Reading infos
INFO
------------------------------------------------
File C:\Users\sangfrois\Desktop\NeuroMod_tests\multi-run.acq contains:
01. GSR-EDA100C-MRI; sampled at 5000.0 Hz
02. PPG100C; sampled at 5000.0 Hz
03. ECG100C; sampled at 5000.0 Hz
04. TTL; sampled at 5000.0 Hz
05. Custom, HLT100C - A 6; sampled at 5000.0 Hz
------------------------------------------------

INFO       saving channel plot to C:\Users\sangfrois\git\phys2bids\code\conversion\multi-run.png
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints according to the std_thr method is 2453. The computed threshold is 0.3033
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
WARNING
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
phys2bids will split the input file according to the given -tr and -ntp arguments
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2453
INFO       Checking number of timepoints
WARNING    Found 2047 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 0.0 seconds and 1328.1316 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2047
INFO       Checking number of timepoints
WARNING    Found 1641 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 1346.728 seconds and 1969.668 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 1641
INFO       Checking number of timepoints
WARNING    Found 1235 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 2006.4526 seconds and 2629.3926 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 1235
INFO       Checking number of timepoints
WARNING    Found 1232 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 2624.3546 seconds and 2646.8246 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 1232
INFO       Checking number of timepoints
WARNING    Found 822 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 3320.5608 seconds and 3949.4608 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 822
INFO       Checking number of timepoints
WARNING    Found 412 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 3962.286 seconds and 4591.186 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 412
INFO       Checking number of timepoints
WARNING    Found 2 timepoints more than expected!
Assuming extra timepoints are at the end (try again with a more liberal thr)
INFO
--------------------------------------------------------------
Slicing between 4606.9372 seconds and 5235.8372 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO
--------------------------------------------------------------
Slicing between 5230.9196 seconds and 5242.4486 seconds
--------------------------------------------------------------
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 406
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 406
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 406
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 3
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 410
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 410
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 410
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Trigger channel name is not in our trigger channel name alias list. Please make sure you choose the proper channel.
INFO       Counting trigger points
INFO       The trigger is in channel 4
INFO       The number of timepoints found with the manual threshold of 0.3033 is 2
INFO       Checking number of timepoints
INFO       Found just the right amount of timepoints!
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Plot trigger
INFO       Found 8 different scans in input!
INFO       Preparing 8 output files.
INFO       Exporting files for run 1 freq 5000.0
```

### The code behaves properly
This needs to be merged. Changes are parcimonious and well structure. Though, I have a little comment :  

``INFO       The number of timepoints according to the std_thr method is 2453. The computed threshold is 0.3033``

What is std_thr method actually ? Does it stand for standard deviation ? If yes, I imagine is related to the current trigger deteciton method in main version. So, I suggest we change the log. Just hit 👍 if you want me to do it!

sangfrois commented:
Also, might I point out that the Y axis on the visualization does not have the same shape as in the source file. On the source file, the trigger values range from 0 to 5 (as you can see on previous graphs), but our trigger plot gives us values ranging from 0 to 1.

![image](https://user-images.githubusercontent.com/38385719/115933255-7f065500-a45c-11eb-9334-44076749b10b.png)

Is this normal ?

RayStick commented:
@sangfrois thanks!

smoia commented:
I guess we can merge this PR at this point - @RayStick @sangfrois are we happy with the results?

sangfrois commented:
@smoia Totally! :)

RayStick commented:
I'm sorry I haven't done any testing yet :pensive:
If it works for @sangfrois I'd say merge it, but I can still test down the line
